src/dry-runner.js

```diff
@@ -0,0 +1,136 @@
+import childProcess from "child_process";
+import {Promise} from "es6-promise";
+import fileUtils from "./file-utils";
+import fs from "fs";
+import logger from "./logger";
+import mockfs from "mock-fs";
+import postpublish from "./postpublish";
+import postversion from "./postversion";
+
+
+const dryRunner = {
+  afterDryRun() {
+    logger.disable();
+    dryRunner.restoreExec();
+    dryRunner.restoreFileSystem();
+
+    return Promise.resolve();
+  },
+
+  beforeDryRun() {
+    logger.enable();
+    dryRunner.patchExec();
+    logger.info("Validating configuration...");
+
+    return fileUtils
+      .readPackage()
+      .then(dryRunner.validatePackage)
+      .then(dryRunner.validateFiles)
+      .then((result) => {
+        dryRunner.patchFileSystem(result.json, result.files);
+
+        return Promise.resolve();
+      });
+  },
+
+  exec: {
```

Comment:
Let's refactor to not patch exec. One idea instead would be to pass an exec function _in_ to the runner that meets the node `exec` signature. Use real node one for real, use a mock one here for dry run.